File: Harden-Windows-Security Module/Main files/C#/CimInstances/BitLocker-EnableKeyProtectors.cs
// MIT License
//
// Copyright (c) 2023-Present - Violet Hansen - (aka HotCakeX on GitHub) - Email Address: spynetgirl@outlook.com
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// See here for more information: https://github.com/HotCakeX/Harden-Windows-Security/blob/main/LICENSE
//

using System;
using System.Globalization;
using System.Management;

namespace HardenWindowsSecurity;

internal partial class BitLocker
{
	/// <summary>
	/// Enables the key protectors of an encrypted volume, doesn't decrypt or encrypt the drive.
	/// The drive can remain encrypted and you use Suspend-BitLocker cmdlet to turn the protection off.
	/// After using this method, the "Protection Status" will be on.
	/// Same as Resume-BitLocker PowerShell cmdlet.
	/// This method must run at the end of the operation when turning on (enabling) BitLocker for the OS drive when it's fully decrypted and has no key protectors.
	/// It can run on a drive where key protectors are already enabled, won't change anything.
	/// https://learn.microsoft.com/en-us/windows/win32/secprov/enablekeyprotectors-win32-encryptablevolume
	/// </summary>
	/// <param name="DriveLetter"></param>
	internal static void EnableKeyProtectors(string DriveLetter)
	{

		// First get the volume info based on the drive letter
		ManagementObject VolumeInfo = GetVolumeFromLetter(DriveLetter);

		// Invoke the method to enable the key protectors
		// The first null indicates that the method does not require any input parameters.
		// The second null indicates that there are no special options being passed for this WMI operation.
		ManagementBaseObject KeyProtectorEnablementResult = VolumeInfo.InvokeMethod("EnableKeyProtectors", null, null);


		#region Output handling
		uint? KeyProtectorEnablementResultCode = null;

		if (KeyProtectorEnablementResult is not null)
		{
			KeyProtectorEnablementResultCode = Convert.ToUInt32(KeyProtectorEnablementResult["ReturnValue"], CultureInfo.InvariantCulture);
		}

		if (KeyProtectorEnablementResultCode is 0)
		{
			Logger.LogMessage($"Successfully enabled the key protectors of the drive {DriveLetter}.", LogTypeIntel.Information);
		}
		else
		{
			HResultHelper.HandleHresultAndLog(KeyProtectorEnablementResultCode);
			return;
		}
		#endregion
	}
}

File: AppControl Manager/eXclude/ComManager/BitLocker/BitLockerEnableKeyProtectors.cpp
#include "BitLockerEnableKeyProtectors.h"
#include "BitLockerManager.h"
#include "..\ComHelpers.h"
#include "..\Globals.h"
#include "..\StringUtilities.h"
#include <Wbemidl.h>
#include <comdef.h>
#include <sstream>
#include <iostream>

using namespace std;

namespace BitLocker {

	// Enables (Resumes) the key protectors of an encrypted volume, doesn't decrypt or encrypt the drive.
	// The drive can remain encrypted and you use Suspend-BitLocker cmdlet to turn the protection off.
	// After using this method, the "Protection Status" will be on.
	// Same as Resume-BitLocker PowerShell cmdlet.
	// This method must run at the end of the operation when turning on (enabling) BitLocker for the OS drive when it's fully decrypted and has no key protectors.
	// It can run on a drive where key protectors are already enabled, won't change anything.
	// https://learn.microsoft.com/windows/win32/secprov/enablekeyprotectors-win32-encryptablevolume
	[[nodiscard]] bool EnableKeyProtectors(const wchar_t* driveLetter)
	{
		// Reset last error buffer for this public API call.
		ClearLastErrorMsg();

		// Basic validation.
		if (!driveLetter || wcslen(driveLetter) != 2 || driveLetter[1] != L':')
		{
			SetLastErrorMsg(L"DriveLetter must be in the form L\"C:\"");
			return false;
		}

		// Connect to BitLocker WMI namespace.
		IWbemLocator* pLoc = nullptr;
		IWbemServices* pSvc = nullptr;
		bool didInitCOM = false;
		if (!ConnectToWmiNamespace(WmiNamespace, &pLoc, &pSvc, didInitCOM) || !pSvc)
		{
			// Error already populated by helper.
			if (pLoc) pLoc->Release();
			if (pSvc) pSvc->Release();
			return false;
		}

		// Resolve instance path for the specified drive.
		wstring instancePath;
		if (!FindVolumeInstancePath(pSvc, driveLetter, instancePath))
		{
			SetLastErrorMsg(L"Failed to locate the BitLocker volume instance path.");
			pSvc->Release();
			pLoc->Release();
			if (!g_skipCOMInit && didInitCOM) CoUninitialize();
			return false;
		}

		// Invoke EnableKeyProtectors (no input parameters).
		IWbemClassObject* pOut = nullptr;
		HRESULT hr = pSvc->ExecMethod(_bstr_t(instancePath.c_str()),
			_bstr_t(L"EnableKeyProtectors"),
			0,
			nullptr,
			nullptr,
			&pOut,
			nullptr);

		if (FAILED(hr) || !pOut)
		{
			SetLastErrorMsg(L"ExecMethod EnableKeyProtectors failed.");
			if (pOut) pOut->Release();
			pSvc->Release();
			pLoc->Release();
			if (!g_skipCOMInit && didInitCOM) CoUninitialize();
			return false;
		}

		// Extract ReturnValue.
		unsigned long rv = static_cast<unsigned long>(-1);
		VARIANT vRet;
		VariantInit(&vRet);
		if (SUCCEEDED(pOut->Get(_bstr_t(L"ReturnValue"), 0, &vRet, nullptr, nullptr)))
		{
			if (vRet.vt == VT_I4)
				rv = static_cast<unsigned long>(vRet.lVal);
			else if (vRet.vt == VT_UI4)
				rv = vRet.ulVal;
		}
		VariantClear(&vRet);

		// Release output object.
		pOut->Release();

		bool success = false;
		if (rv == 0)
		{
			LogOut(L"Successfully enabled the key protectors of the drive ", driveLetter, L".");
			success = true;
		}
		else
		{
			wstringstream ss;
			ss << L"EnableKeyProtectors failed " << FormatReturnCode(rv);
			SetLastErrorMsg(ss.str());
			LogErr(ss.str().c_str());
			success = false;
		}

		// Cleanup COM objects.
		pSvc->Release();
		pLoc->Release();
		if (!g_skipCOMInit && didInitCOM) CoUninitialize();

		return success;
	}
}
